cpp/src/cones.cpp

```diff
@@ -354,3 +354,88 @@ LinearOperator dprojection(const Vector &x, const std::vector<Cone> &cones,
   }
   return block_diag(lin_ops);
 }
+
+void _op_into_dense(MatrixRef &D_block, const LinearOperator &D_op) {
+  int size = D_block.rows();
+  Vector v = Vector::Zero(size);
+  for (int i = 0; i < size; ++i) {
+    v[i] = 1.;
+    D_block.row(i) = D_op.matvec(v);
+    v[i] = 0.;
+  }
+}
+
+int _get_D_size(const std::vector<Cone> &cones) {
+  int offset = 0;
+  for (const Cone &cone : cones) {
+    const ConeType &type = cone.type;
+    const std::vector<int> &sizes = cone.sizes;
+    if (std::accumulate(sizes.begin(), sizes.end(), 0) == 0) {
+      continue;
+    }
+    for (int size : sizes) {
+      if (type == PSD) {
+        size = vectorized_psd_size(size);
+      } else if (type == EXP) {
+        size *= 3;
+      }
+      offset += size;
+    }
+  }
+  return offset;
+}
+
+void _dprojection_pos_dense(MatrixRef &D_block, const Vector &x) {
+  const Array sign = x.cwiseSign();
+  D_block.diagonal() << (0.5 * (sign + 1));
```

Comment:
We can't do that verbatim:

```C++
cpp/src/cones.cpp: In function ‘void _dprojection_pos_dense(MatrixRef&, const Vector&)’:        cpp/src/cones.cpp:391:28: error: no match for ‘operator>=’ (operand types are ‘const Vector {aka const Eigen::Matrix<double, -1, 1>}’ and ‘int’)                                                   D_block.diagonal() << (x >= 0);
                          ~~^~~~
```

But we could maybe create a `gt` function like in `deriv.cpp` but for a `Vector` x and then use that here, and in `_dprojection_pos`. This will incorrectly give `0.5` if `x` is exactly zero, so perhaps we should change this? Ok with you @akshayka?

Reply:
Oh shoot, good catch, yes please fix this.

Reply:
What should we do when it is exactly zero? The gradient is undefined there. Random number between 0 and 1? 0? 1/2? 1?

Reply:
newton_admm code is doing `1(x>=0)`: https://github.com/locuslab/newton_admm/blob/master/newton_admm/cones.py#L23

Reply:
Anything you said should be a valid subgradient though, so shall we just stick with `0`?

Reply:
Or maybe 0.5 is ok too? Is pretty convenient to just keep the componentwise sign operation. Seems somewhat arbitrary of a choice although maybe there's a case where it matters, what do you all think?

Reply:
I think it’s fine to leave as is